state/tree.go

```diff
@@ -243,42 +243,47 @@ func (t *tree) debugPointer(ps []*hamt.Pointer) {
 	}
 }
 
+// GetAllActorsResult is the struct returned via a channel by the GetAllActors
+// method. This struct contains only an address string and the actor itself.
+type GetAllActorsResult struct {
+	Address string
+	Actor   *actor.Actor
+}
+
 // GetAllActors returns a slice of all actors in the StateTree, t.
-func GetAllActors(t Tree) ([]string, []*actor.Actor) {
+func GetAllActors(t Tree) <-chan GetAllActorsResult {
 	st := t.(*tree)
-
-	return st.getActorsFromPointers(st.root.Pointers)
+	out := make(chan GetAllActorsResult)
+	go func() {
+		st.getActorsFromPointers(out, st.root.Pointers)
+		close(out)
+	}()
+	return out
 }
 
-// GetAllActorsFromStoreFunc is a function with the signature of GetAllActorsFromStore
-type GetAllActorsFromStoreFunc = func(context.Context, *hamt.CborIpldStore, cid.Cid) ([]string, []*actor.Actor, error)
-
-// GetAllActorsFunc is a function with the signature of GetAllActors
-type GetAllActorsFunc = func(t Tree) ([]string, []*actor.Actor)
-
 // GetAllActorsFromStore loads a StateTree and returns arrays of addresses and their corresponding actors.
 // Third returned value is any error that occurred when loading.
-func GetAllActorsFromStore(ctx context.Context, store *hamt.CborIpldStore, stateRoot cid.Cid) ([]string, []*actor.Actor, error) {
+func GetAllActorsFromStore(ctx context.Context, store *hamt.CborIpldStore, stateRoot cid.Cid) (<-chan GetAllActorsResult, error) {
 	st, err := LoadStateTree(ctx, store, stateRoot, nil)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-
-	addrs, actors := GetAllActors(st)
-	return addrs, actors, nil
+	return GetAllActors(st), nil
```

Comment:
blocking: seems like getallactors needs to respect the context else it cannot be canceled